lib/types/rules/ecmascript-6.d.ts

```diff
@@ -291,23 +313,39 @@ export interface ECMAScript6 extends Linter.RulesRecord {
         [
             ...Array<
                 | string
-                | {
-                    name: string;
-                    importNames?: string[] | undefined;
-                    message?: string | undefined;
-                }
+                | ({
+                      name: string;
+                      message?: string;
+                      importNames?: string[];
+                      allowImportNames?: string[];
+                  } & (
+                      | { importNames?: string[]; allowImportNames?: never }
+                      | { allowImportNames?: string[]; importNames?: never }
+                  ))
                 | Partial<{
-                    paths: Array<
-                        | string
-                        | {
-                            name: string;
-                            importNames?: string[] | undefined;
-                            message?: string | undefined;
-                        }
-                    >;
-                    patterns: string[];
-                }>
-            >,
+                      paths: Array<
+                          | string
+                          | ({
+                                name: string;
+                                message?: string;
+                                importNames?: string[];
+                                allowImportNames?: string[];
+                            } & (
+                                | {
+                                      importNames?: string[];
+                                      allowImportNames?: never;
+                                  }
+                                | {
+                                      allowImportNames?: string[];
+                                      importNames?: never;
+                                  }
+                            ))
+                      >;
+                      patterns: Array<
+                          string | ValidNoRestrictedImportPatternOptions
+                      >;
+                  }>
+            >
```

Comment:
[Style] This is getting to be pretty deep & complex 😅. Suggestion: extract out the reused portions into named interfaces, such as the two `{ importNames?: ...; allowImportNames?: ... }` shapes?

For example, the `ValidNoRestrictedImportPatternOptions` type could look something like:

```ts
type ValidNoRestrictedImportPatternOptions =
  NoRestrictedImportPatternCommonOptions &
    EitherGroupOrRegEx &
    EitherNameSpecifiers;
```

It might also make sense to write a utility type to generate the _"union of objects, where each only allows one property"_ types. [`type-fest`](https://github.com/sindresorhus/type-fest)'s [`MergeExclusive`](https://github.com/sindresorhus/type-fest/blob/3d473d1f8b0ec5820e1f3aaf7965b68784423778/source/merge-exclusive.d.ts) does that for just two. [Here's a playground for a type that allows arbitrary objects](https://www.typescriptlang.org/play/?exactOptionalPropertyTypes=true#code/C4TwDgpgBA6glsAFgewK7ADwBUA0UDSEIAggDanIDuEAJlBAB7AQB2NAzlANZHIBmULAD4oAXigBvAL5QAZJIBQUKAG1CIAJIsWEAE5Q4LKAFEGAY1KoaEDDxD9BedWQrUaQgLoB+AFxQdAG56ANwKUqEKoJBQAApwZlzEbNhO9EysHNy8AsJikjLyEkqqMQZG+B5+WCoxHqHhCpHg0HEJSTQAInDsAIbkVADySHrsKQRELlS0acxsnHYOuaLF8q2JybjjJP1uQiuwCCjoY847tEIRUdAAqixwyCwAYsi6AyykIG8Q6tgi4tJyRTKNRlLL2HKVWLxdadbp9VxDRAjE4XMIqBYQy7NKAANT6cBoADlkAAlCDsYC6eLMGgaAC2YBewBiPWAzF0LAGYGA9xYnH+BSgt15z1e70+Oh+RWU8Ko9MZumAhJ6dIgLLZehYfgpVJYAHMVHVirLKPKmcrVextZTDAajco4AzzSq1az2VqoDrbYbQg6nYqLeTrbq7fVUQprBYerpoGYHhSoMhubyrbj8UTSeSbWYaWbFer3VyefGInABAAKJPFvkAOkdCqVLvYAEogYnkyXigB6LtQAB6XjCQA), with `{} &`s added to get TypeScript to print the type nicely after narrowing. 🤷 that might be a bit heavyweight for just this one rule's types though.

Reply:
>that might be a bit heavyweight for just this one rule's types though.

Yeah, will look into it if there are more rule types like this.